### DISCUSSION TITLE
E310 support

### DISCUSSION THREAD

**Discussion by alphafox02** [open-sdr/openwifi#333] [General]
I'm curious, price wise it's much more expensive then something like the E310/200, but could OpenWiFi run on the E310/320? 
https://www.ettus.com/all-products/e310/

**Comment by JiaoXianjun** [comment]
We never try ettus e310/320. In principle, the porting could be done, because the architecture is similar to the those platforms we support.

**Reply by alphafox02** [reply]
That could be very interesting but maybe not widely adapted? But still, would be great. You would likely need hardware on your end though?

**Reply by JiaoXianjun** [reply]
We have lots of hardwares and supported many platforms. Some of them are even much cheaper than ettus e310. So currently there is not strong motivation from our side regarding porting openwifi to ettus hardware. But of course we would be glad to answer questions about our current design on current platforms if some one from the community wants to port the openwifi to ettus e310.

**Reply by alphafox02** [reply]
I’d like to try,  but I think it may be a little above my level at this point. If you do happen to make it available on your end I will happily test with the e310 on my end.

**Comment by alphafox02** [comment]
I would like to try to port to the E310, but wouldn't know where to begin. Seems I would need to begin here?

https://github.com/open-sdr/openwifi#porting-guide

**Comment by Fr0stByte00** [comment]
Attempted to do this and have run into numerous road blocks. For starters the open-wifi repo depends on Vivado 2023.2 and then your open-wifi-hw repo depends on 2021.1. These are massive software packages and I am not sure why the hardware part doesn't use the latest one. I used HDL's [usrpe31x](https://github.com/analogdevicesinc/hdl/tree/main/projects/usrpe31x) code to compile a system_top.xsa file with relative ease and that was about all that supports the E310. Generating the zynq BOOT.BIN I followed this [guide](https://wiki.analog.com/resources/tools-software/linux-software/build-the-zynq-boot-image) but then had to generate the u-boot.elf. Of course that had it's own [instructions](https://xilinx-wiki.atlassian.net/wiki/spaces/A/pages/18841973/Build+U-Boot) and the Xilinx U-Boot repo makes no references to the Zynq 7020 SoC so I am assuming I use the Zynq ZC702 for that. Once I generated that I was able to use the script given in the boot image resource to generate the BOOT.BIN. As for the device tree blob - that I just pulled off my E310 image directly. I am not sure if that would work but it is beyond my skill to manually generate that myself. So now all that's left is to generate the `system_top.bit.bin`. And ofc that's where they direct you to the [open-wifi-hw](https://github.com/open-sdr/openwifi-hw/tree/master)  and it depends on Vivado 2021.1. Why does all of the rest of the work require 2023.2 and this is 2021.1?? This repo doesn't seem to allow someone to port their own board in - I analyzed the shell scripts in it and the `prepare_adi_board_ip.sh` one checks to make sure it's one of the supported boards on your list. Am I missing something? From what I could tell if I could generate a `BOOT.BIN`, a `system_top.bit.bin`, and a `devicetree.dtb` for my board and then port that onto an open_wifi image then I could get it working for the E310. I am not sure how to convert my system_top.bit to a bin or really properly generate the FPGA bit stream using your open-wifi-hw repo, it seems fully dependent on a user using one of your supported boards. From what I can tell the in the open-wifi-hw repo each board folder contains the generated `ip_repo_gen.tcl` and the `openwifi.tcl`. The `ip_repo_gen.tcl` seems to be the same in all of them but the other is generated through vivado. What files should I port/compile into vivado to do that? After compiling the HDL code for the usrpe31x I have all of the files in the `src` folder. so I think I am close and could edit the code to add in my board. Any clarification is appreciated.

**Reply by Fr0stByte00** [reply]
The main question is - how do I generate that openwifi.tcl file for my board? HDL supports the usrpe31x series so I feel like this should not be too bad to tie that into the open-wifi-hw repo. **EDIT** I was able to edit the file structure in the open-wifi repo to add in my usrpe31x files to run the [boot_bin_gen.sh](https://github.com/open-sdr/openwifi/blob/master/user_space/boot_bin_gen.sh) script. I uploaded the BOOT.BIN, devicetree.dtb, and system_top.bit.bin to an open wifi image and then pulled them into the BOOT base folder. The E310 would not even boot up so I suspect their OS which is an alchemy linux image just isn't compatible with open wifi's kuiper image. Not totally sure but I am not going to try and somehow integrate open wifi specifics onto their base image

**Reply by Switcherag** [reply]
Thanks for the previous job,I will give it a try if you want to share some insight or anything that can help you are welcome

**Comment by alphafox02** [comment]
@JiaoXianjun Wondered perhaps if you have thoughts on recent comments? Would love to see another hardware device supported by OpenWiFi. It seems like there is interest by someone in porting.
